File: Code/Johan_Mendoza/automata-1-Mendoza.cpp
#include <iostream>
#include <vector>

using namespace std;


/*
S0 = 0
S1 = 1
S2 = 2
S3 = 3

State = -1 ---> input sequence not legal for state

Accept input in form a,a,b,c for array, or a single value.
*/


int nextState(int current_state, char inputs){

    if (inputs == 'a' && current_state == 0){
        return 2;
    } else if (inputs == 'a' && current_state == 2){
        return 0;
    } else if (inputs == 'b' && current_state == 0){
        return 1;
    } else if (inputs == 'b' && current_state == 1){
        return  1;
    } else if (inputs == 'c' && current_state == 0){
        return 3;
    } else if (inputs == 'c' && current_state == 1){
        return  3;
    } else {
        return -1;
    }


}



//* The function split, split an char array.
vector<char> split (char *arr, int arr_len, char splitby){
    vector<char> list;

    for(int i = 0; i < arr_len; i++){
        if (arr[i] != splitby) {
            auto it = list.insert(list.end(), arr[i]);
        }
    }
    return list;
}

int main()
{
	char array[10];
	char i;
	int state;
	int current_state = 0;

	vector<char> list;

    cout << "Type 'x' to quit the program" << endl;
    cout << "Accept input in format = a,a,b,c or a single value" << endl;

    //*std::cout << "Stato iniziale:" << current_state << endl;
	while(true)
	{
        cout << "Enter sequence = ";
        cin >> array;

        list = split(array,(unsigned)strlen(array),',');

	    for (int c = 0; c < list.size(); c++){
	        i = list[c];
            cout << "Current state: S" << current_state << endl;

            if (i == 'x'){
                cout << "Exited" << endl;
                return 0;
            }

            state = nextState(current_state,i);

            if(state == -1){
                cout << "Input sequence (" << i << ") not legal for state S" << current_state << endl;
                return -1;
            } else if(state == 3) {
                cout << "we're in final state: S3" << endl;
                return 0;
            } else {
                cout << "Next state: S" << state << endl;
                current_state = state;
            }

	    }

	}

}

File: Code/Johan_Mendoza/automata-1-Mendoza.py
"""
  Computes next state. Returns the following

    -1 : input sequence not legal
    -2 : we're in final state
"""

import sys

def nextstate(current_state,inputs):

    if inputs == "a" and current_state == "S0":
        current_state = "S2"
    elif inputs == "a" and current_state == "S2":
        current_state = "S0"
    elif inputs == "b" and current_state == "S0":
        current_state = "S1"
    elif inputs == "b" and current_state == "S1":
        current_state = "S1"
    elif inputs == "c" and current_state == "S0":
        current_state = "S3"
        return -2
    elif inputs == "c" and current_state == "S1":
        current_state = "S3"
        return -2
    else:
        return -1
    return current_state

if __name__ == '__main__':
    #print(sys.argv[1])

    #get input from command line
    if len(sys.argv) < 2:
        print("Error, insert an input")
        sys.exit()
    elif len(sys.argv) > 2:
        print("Error, expected only 1 argoment/array")
        sys.exit()
    else:
        inputs = list(map(str, sys.argv[1].rstrip().split(",")))

    #inputs = ['a','a','b','c']

    current_state = "S0" #di default

    for i in inputs:

        print("Current state:", current_state)

        stato = nextstate(current_state,i)

        if stato == -1:
            print("input sequence ({inputs}) not legal for state {stato}".format(inputs=i, stato=current_state))
            break

        elif stato == -2:
            print("we're in final state: S3")
            break
        else:

            print("Next state:",stato)

            current_state = stato

    print("Done")